craft_store/http_client.py

```diff
@@ -0,0 +1,152 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Craft Store HTTPClient."""
+
+import logging
+import os
+from typing import Dict, Optional
+
+import requests
+from requests.adapters import HTTPAdapter, Retry
+
+from . import errors
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPClient:
+    """Generic HTTP Client to communicate with Canonical's Developer Gateway.
+
+    This client has a requests like interface, it creates a requests.Session
+    on initialization to handle retries over HTTP and HTTPS requests.
+
+    The default number of retries is set to 5 and can be overridden through
+    the ``CRAFT_STORE_RETRIES`` environment variable.
+
+    The backoff factor has a default of 2 and can be overridden with the
+    ``CRAFT_STORE_BACKOFF`` environment variable.
+
+    Retries are done for the following return codes: ``104``, ``500``, ``502``,
+    ``503`` and ``504``.
+
+    :ivar user_agent: User-Agent header to identify the client.
+    """
+
+    def __init__(self, *, user_agent: str) -> None:
+        """Initialize and HTTPClient with a given user_agent.
+
+        :param user_agent: User-Agent header to identify the client.
+        """
+        self._session = requests.Session()
+        self.user_agent = user_agent
+
+        try:
+            total = int(os.environ.get("CRAFT_STORE_RETRIES", 10))
+        except ValueError:
+            total = 10
+            logger.debug(
+                "CRAFT_STORE_RETRIES is not set to an integer, using default of %r.",
+                total,
+            )
+
+        try:
+            backoff = int(os.environ.get("CRAFT_STORE_BACKOFF", 2))
+        except ValueError:
+            backoff = 2
```

Comment:
A total retries of 10 and backoff of 2 will lead to a max delay of 17 minutes, which is not acceptable. Also, first retry should be much faster, as normally those are super transient glitches.

I suggest using this values: `backoff=.2` and `total=8`; this allows a super fast first retry, and a total timeout just smaller than a minute.

Reply:
I reworked the retry logic value retrieval into a common method and set these defaults as constants to be used on https://github.com/canonical/craft-store/pull/1/commits/8e6db4219cd9eb1a653075e7809d8312b864db60